File: day4/part2.py
#!/usr/bin/env python

from __future__ import annotations
from typing import List, Tuple


def load_grid(path: str) -> List[List[str]]:
    """
    Load non-empty lines from `path` into 2D list of characters.
    """
    grid: List[List[str]] = []
    with open(path, "r", encoding="utf-8") as f:
        for raw in f:
            line = raw.strip("\n")
            if line.strip():
                grid.append(list(line))
    return grid


def find_accessible(grid: List[List[str]]) -> List[Tuple[int, int]]:
    """
    Return a list of coordinates (r, c) of '@' cells that have fewer
    than 4 adjacent '@' neighbors.
    """
    if not grid:
        return []

    h = len(grid)
    dirs: List[Tuple[int, int]] = [
        (-1, -1),
        (-1, 0),
        (-1, 1),
        (0, -1),
        (0, 1),
        (1, -1),
        (1, 0),
        (1, 1),
    ]

    accessible: List[Tuple[int, int]] = []

    for r in range(h):
        row_len = len(grid[r])
        for c in range(row_len):
            if grid[r][c] != "@":
                continue
            adj = 0
            for dr, dc in dirs:
                nr, nc = r + dr, c + dc
                if 0 <= nr < h and 0 <= nc < len(grid[nr]) and grid[nr][nc] == "@":
                    adj += 1
                    if adj >= 4:
                        break
            if adj < 4:
                accessible.append((r, c))

    return accessible


def remove_until_stable(grid: List[List[str]]) -> int:
    """
    Repeatedly remove all accessible '@' (set to '.') until no more can be removed.
    Return total removed count.
    """
    total_removed = 0
    while True:
        to_remove = find_accessible(grid)
        if not to_remove:
            break
        for r, c in to_remove:
            if grid[r][c] == "@":
                grid[r][c] = "."
                total_removed += 1
    return total_removed


def main() -> None:
    grid = load_grid("input.txt")
    removed = remove_until_stable(grid)
    print(removed)


if __name__ == "__main__":
    main()

File: day4/part2.ts
import * as fs from "fs";

type Grid = string[][];

/**
 * Load grid from input.txt, ignoring empty lines.
 */
function loadGrid(path: string): Grid {
  const raw = fs.readFileSync(path, "utf8");
  return raw
    .split(/\r?\n/)
    .filter((ln) => ln.trim().length > 0)
    .map((ln) => ln.split(""));
}

/**
 * Find all accessible '@' cells (r,c) with fewer than 4 adjacent '@'.
 */
function findAccessible(grid: Grid): Array<[number, number]> {
  if (grid.length === 0) return [];

  const dirs: Array<[number, number]> = [
    [-1, -1],
    [-1, 0],
    [-1, 1],
    [0, -1],
    [0, 1],
    [1, -1],
    [1, 0],
    [1, 1],
  ];

  const result: Array<[number, number]> = [];

  for (let r = 0; r < grid.length; r++) {
    for (let c = 0; c < grid[r].length; c++) {
      if (grid[r][c] !== "@") continue;
      let adj = 0;
      for (const [dr, dc] of dirs) {
        const nr = r + dr;
        const nc = c + dc;
        if (nr >= 0 && nr < grid.length && nc >= 0 && nc < grid[nr].length) {
          if (grid[nr][nc] === "@") {
            adj++;
            if (adj >= 4) break; // short-circuit
          }
        }
      }
      if (adj < 4) result.push([r, c]);
    }
  }
  return result;
}

/**
 * Repeatedly remove accssible rolls until stable. Return total removed.
 */
function removeUntilStable(grid: Grid): number {
  let total = 0;
  while (true) {
    const toRemove = findAccessible(grid);
    if (toRemove.length === 0) break;
    for (const [r, c] of toRemove) {
      if (grid[r][c] === "@") {
        grid[r][c] = ".";
        total++;
      }
    }
  }
  return total;
}

function main(): void {
  const grid = loadGrid("input.txt");
  console.log(removeUntilStable(grid));
}

main();
